### PR TITLE
logging: rpc: log history support

### PR DESCRIPTION
**Pull request by Damian-Nordic**
Add initial code for supporting the log history, that is, the mode of nRF RPC logging library in which log messages are stored in a ring buffer, and nRF RPC commands can be used for configuring and fetching the log history.

The rationale for supporting this mode is to reduce the nRF RPC traffic so that the bandwidth is available for higher priority tasks, while preserving the possibility to analyse the logs when an issue occurs.

This commit adds the support for saving the log messages in the ring buffer, fetching the log history, and setting the maximum level of saved messages.

### CODE REVIEW

**Review by alxelax — COMMENTED** [review]
LGTM just a couple of questions.

**Review by canisLupus1313 — APPROVED** [review]

File: samples/nrf_rpc/protocols_serialization/client/src/log_rpc_shell.c [hunk 1]
```diff
@@ -33,6 +33,72 @@ static int cmd_log_rpc_stream_level(const struct shell *sh, size_t argc, char *a
...
+
+	if (rc) {
+		shell_error(sh, "Error: %d", rc);
+		return -ENOEXEC;
+	}
+
+	return 0;
+}
+
+static const struct shell *shell;
```
[the remark this answers is not in the document]
Comment (Damian-Nordic) [reply samples/nrf_rpc/protocols_serialization/client/src/log_rpc_shell.c:58]: Implemented the remaining functionality.

File: subsys/logging/log_backend_rpc.c [hunk 2]
```diff
@@ -350,3 +363,50 @@ static void log_rpc_set_stream_level_handler(const struct nrf_rpc_group *group,
...
+static void log_rpc_fetch_history_handler(const struct nrf_rpc_group *group,
+					  struct nrf_rpc_cbor_ctx *ctx, void *handler_data)
+{
+	if (!nrf_rpc_decoding_done_and_check(group, ctx)) {
+		nrf_rpc_err(-EBADMSG, NRF_RPC_ERR_SRC_RECV, group, LOG_RPC_CMD_FETCH_HISTORY,
+			    NRF_RPC_PACKET_TYPE_CMD);
+		return;
+	}
+
+	/* TODO: implement log upload:
```
[the remark this answers is not in the document]
Comment (Damian-Nordic) [reply subsys/logging/log_backend_rpc.c]: Implemented the remaining functionality instead.

File: subsys/logging/log_forwarder_rpc.c [hunk 3]
```diff
@@ -22,6 +22,8 @@
 
 LOG_MODULE_REGISTER(remote);
 
+static log_rpc_history_handler_t history_handler;
```
[the remark this answers is not in the document]
Comment (Damian-Nordic) [reply subsys/logging/log_forwarder_rpc.c:28]: Implemented the remaining functionality.

**Review by nordicjm — APPROVED** [review]

File: subsys/logging/log_forwarder_rpc.c [hunk 4]
```diff
@@ -72,6 +77,94 @@ int log_rpc_set_stream_level(enum log_rpc_level level)
 	return 0;
 }
 
+int log_rpc_set_history_level(enum log_rpc_level level)
```
Comment (konradderda) [inline subsys/logging/log_forwarder_rpc.c:80]: This function will never return any other value than 0, so maybe use void?

File: subsys/logging/log_forwarder_rpc.c [hunk 5]
```diff
@@ -72,6 +77,94 @@ int log_rpc_set_stream_level(enum log_rpc_level level)
...
+
+	NRF_RPC_CBOR_ALLOC(&log_rpc_group, ctx, 1 + sizeof(level));
+	nrf_rpc_encode_uint(&ctx, level);
+	nrf_rpc_cbor_cmd_no_err(&log_rpc_group, LOG_RPC_CMD_SET_HISTORY_LEVEL, &ctx,
+				nrf_rpc_rsp_decode_void, NULL);
+
+	return 0;
+}
+
+int log_rpc_fetch_history(log_rpc_history_handler_t handler)
```
Comment (konradderda) [inline subsys/logging/log_forwarder_rpc.c:92]: Function always returns true - maybe use `void`?

**Review by konradderda — APPROVED** [review]
Looks good, found two nits which you can apply or ignore :)

**Pull request merged**